oxidation/libparsec/crates/types/src/regex.rs

```diff
@@ -12,43 +18,132 @@ fn escape_globing_pattern(string: &str) -> String {
 }
 
 impl Regex {
-    pub fn from_pattern(pattern: &str) -> Result<Self, fnmatch_regex::error::Error> {
-        let escaped_str = escape_globing_pattern(pattern);
-        Ok(Regex(fnmatch_regex::glob_to_regex(&escaped_str)?))
+    /// Returns a regex which is built from a file that contains shell like patterns
+    pub fn from_file(file_path: &Path) -> RegexResult {
+        let file_content =
+            fs::read_to_string(file_path).map_err(|err| RegexError::PatternFileIOError { err })?;
+
+        Ok(Self(
+            file_content
+                .lines()
+                .into_iter()
+                .map(str::trim)
+                .filter(|l| *l != "\n" && !l.starts_with('#'))
+                .map(|l| {
+                    fnmatch_regex::glob_to_regex(&escape_globing_pattern(l))
+                        .map_err(|err| RegexError::GlobPatternError { err })
+                })
+                .collect::<Result<Vec<regex::Regex>, RegexError>>()?,
+        ))
     }
 
-    pub fn from_regex_str(regex_str: &str) -> Result<Self, regex::Error> {
-        Ok(Regex(regex::Regex::new(regex_str)?))
+    /// Returns a regex which is an union of all regexes from `raw_regexes` slice parameter
+    pub fn from_raw_regexes(raw_regexes: &[&str]) -> RegexResult {
+        Ok(Self(
+            raw_regexes
+                .iter()
+                .map(|l| regex::Regex::new(l).map_err(|err| RegexError::ParseError { err }))
+                .collect::<Result<Vec<regex::Regex>, RegexError>>()?,
+        ))
     }
 
-    pub fn is_match(&self, string: &str) -> bool {
-        self.0.is_match(string)
+    /// Returns a regex from a glob pattern
+    pub fn from_glob_pattern(pattern: &str) -> RegexResult {
+        let escaped_str = escape_globing_pattern(pattern);
+        Ok(Regex(vec![fnmatch_regex::glob_to_regex(&escaped_str)
+            .map_err(|err| RegexError::GlobPatternError { err })?]))
     }
-}
 
-impl AsRef<str> for Regex {
-    fn as_ref(&self) -> &str {
-        self.0.as_str()
+    pub fn from_regex_str(regex_str: &str) -> RegexResult {
+        Self::from_raw_regexes(&[regex_str])
+    }
+
+    pub fn is_match(&self, string: &str) -> bool {
+        self.0.iter().any(|r| r.is_match(string))
     }
 }
 
 impl PartialEq for Regex {
     fn eq(&self, other: &Self) -> bool {
-        self.as_ref() == other.as_ref()
+        if self.0.len() == other.0.len() {
+            let set: HashSet<&str> = HashSet::from_iter(self.0.iter().map(|r| r.as_str()));
+
+            other.0.iter().all(|r| set.contains(r.as_str()))
+        } else {
+            false
+        }
     }
 }
 
 impl Display for Regex {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        self.0.fmt(f)
+        for (i, r) in self.0.iter().enumerate() {
+            r.fmt(f)?;
+
+            if i + 1 < self.0.len() {
+                f.write_char('|')?;
+            }
+        }
+
+        Ok(())
```

Comment:
What do you prefer ?
```Rust
writeln!(f, self.0.iter().map(regex::Regex::as_str).collect::<Vec<_>>().join("|"))
```

Reply:
Let's use your version it's shorter :wink: